### ISSUE TITLE
ActiveRel creation fails if a related node is subclassed

### ISSUE THREAD

**Issue by brucek** [neo4jrb/neo4j#1133]
I have a [failing test](https://github.com/brucek/neo4j/blob/a2b80841204667f80f206755ea021ed1d725a853/spec/e2e/active_rel/persistence/query_factory_spec.rb) I'm trying to puzzle out now...

Basically this fails:

    class ParentClass
      include Neo4j::ActiveNode
    end
    
    class FromClass < ParentClass
      has_many :out, :to_classes, rel_class: 'RelClass'
    end
    
    class ToClass
      include Neo4j::ActiveNode
    end
    
    class RelClass
      type 'is_a'
      from_class :FromClass
      to_class :ToClass
    end

with this error:

    Neo4j::Session::CypherError: ParentTestClass_uuid not defined
      "MATCH (to_node) WHERE (ID(to_node) = {to_node_id})
      CREATE (from_node:`FromClass:ParentClass` {uuid: {from_node_FromClass:ParentClass_uuid}})
      CREATE UNIQUE from_node-[rel:is_a {rel_create_props}]->to_node
      SET rel.`created_at` = {setter_rel_created_at}
      RETURN rel, from_node"

It looks like it's a problem with that parameter needing to be quoted maybe to hide the identifier colon(:) from the parser?  But I'm still digging out where exactly the issue originates.  I wonder if this is related to #1128 also?

**Comment by subvertallchris** [comment]
Interesting!

ActiveRel's queries get built using classes defined [here](https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/shared/query_factory.rb). You'll get different results based on the persistence state of the nodes. We can see from [this line](https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/shared/query_factory.rb#L72) that we're just using the Cypher DSL from Neo4j-core, so I think the root cause of this is the way that param names are created from labels. When I run that problem line directly, IRB singles out that `:`.

```
Neo4j::Session::CypherError: ParentClass_uuid not defined (line 1, column 51 (offset: 50))
"CREATE (n:`Student:ParentClass` {uuid: {n_Student:ParentClass_uuid}}) RETURN n"
                                                   ^
```

We should take a look at Neo4j-core, find where it sets params, and make sure it strips invalid characters.

**Comment by subvertallchris** [comment]
Spent 30 mins looking at Neo4j-core and it looks like it might be tricky to fix, but @cheerfulstoic is more familiar with that code than me and might be able to do it faster. As a quick and dirty fix, we can just rewrite the CREATE statement with a String and set params manually.

```ruby
    def create_query
      return match_query if graph_object.persisted?
      base_query.create("(#{identifier}:`#{graph_object.labels_for_create.join('`:`')}` {#{identifier}_create_params})")
        .params("#{identifier}_create_params" => graph_object.props_for_create)
    end
```

I didn't write a test for it so I won't PR it or merge in just yet, have to get off the computer right now (actually 10 minutes ago!) but it appears to solve the problem when I test by hand.

Alternatively, as a workaround, you can create the node and rel in two steps instead of one. The params won't be part of the rel creation query so it won't break.

**Comment by brucek** [comment]
Thanks Chris - hopefully [this fix](https://github.com/neo4jrb/neo4j-core/pull/251) will work in neo4j-core

**Comment by subvertallchris** [comment]
Awesome! Fix released as 6.1.7!

**Comment by cheerfulstoic** [comment]
:dancers:
